### PR TITLE
chore: dev cli support for calling master apis

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

Add `dev cli b|indings` to allow discovering and calling various master endpoints.
There are limitations to what a simple cli interface would allow for in terms of providing arguments, and more so with the current implementation yet as is it should allow making about ~110 different api calls. This is assuming our api annotations are accurate.

## Test Plan

`det dev b -h`
- added automated tests

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by erikwilson — APPROVED** [review]

**Comment by hamidzr** [comment]
I should check this out https://goswagger.io/generate/cli.html

**Review by rb-determined-ai — COMMENTED** [review]
Technically speaking, I think this is really cool.

From a product perspective, I think this is a bad idea.

Can we pull `det dev` out and make it a separate script?

I don't like where this is going, adding more and more developer functionality into a tool for end users.

I especially don't like increasing the discoverability of our v1 REST API, when we all know our v1 REST API is poorly designed.  I don't want to encourage users to use it.  My proposals to formally demote it from public to private status have been rejected, but we don't need to make it easier for end users to discover rest api endpoints.

**Review by rb-determined-ai — APPROVED** [review]
Hamid and I discussed further, and I'd still rather this be made a standalone tool for developers than to ship dev tools as part of our product.  Same goes for all the internal paths of our REST API, which we ship in our swagger-generated docs.

But it is under the `dev` namespace so I'm not going to block this PR.  Also the fact that the `dev` tools have proven useful in a couple of support cases is a convincing argument.

File: harness/determined/cli/dev.py
Comment (hamidzr) [inline harness/determined/cli/dev.py:70]: these don't need to be underscored

**Comment by hamidzr** [comment]
> @[hamidzr](https://github.com/hamidzr) hamidzr requested review from [ashtonG](https://github.com/ashtonG) (assigned from determined-ai/web) and [wes-turner](https://github.com/wes-turner) (assigned from determined-ai/ml-sys) and removed request for https://github.com/orgs/determined-ai/teams/ml-sys [2 minutes ago](https://github.com/determined-ai/determined/pull/7462#event-10013345937)

I don't know where these review requests came from 🤷🏼

**Comment by keita-determined** [comment]
> > @[hamidzr](https://github.com/hamidzr) hamidzr requested review from [ashtonG](https://github.com/ashtonG) (assigned from determined-ai/web) and [wes-turner](https://github.com/wes-turner) (assigned from determined-ai/ml-sys) and removed request for https://github.com/orgs/determined-ai/teams/ml-sys [2 minutes ago](https://github.com/determined-ai/determined/pull/7462#event-10013345937)
> 
> I don't know where these review requests came from 🤷🏼

i think its all from https://github.com/determined-ai/determined/blob/577cebe8a35084ed3e12d8e74a087e154b891396/.github/CODEOWNERS#L1-L27

File: harness/determined/cli/dev.py
Comment (wes-turner) [reply harness/determined/cli/dev.py:71]: I think the inline comments describe the input/output. My recommendation is to put this spec into the docstring instead of into comments.

File: harness/determined/cli/dev.py
Comment (wes-turner) [inline harness/determined/cli/dev.py:206]: I'll drop more comments as I see them, but just in general I think it's more idiomatic in Python to favor expressive names over brief names a little more than this. Could you please name `d` (and later `v`) with what it stands for?

Comment (wes-turner) [inline harness/determined/cli/dev.py:206]: I think it's just in general easier to test and reason about pure functions than about impure ones. With printing, I think it can also help maintain the boundary between view and controller. What do you think about changing the interface here to return a string that the controller prints?

File: harness/determined/cli/dev.py
Comment (wes-turner) [inline harness/determined/cli/dev.py:239]: Thanks for being opinionated. The interface is a little weird. If it was my own, I'd probably spit out an error message and exit, but you're already using this and you know what works for you so whatever you like is fine. If you do decide to keep it, could you clean up the language a bit to make it more professional software?

File: harness/determined/cli/dev.py
Comment (wes-turner) [inline harness/determined/cli/dev.py:253]: This is nice.

File: harness/determined/cli/dev.py
Comment (wes-turner) [inline harness/determined/cli/dev.py:252]: I'm not sure it's ever right to catch all exception types. What are you trying to be inclusive to catch?

File: harness/determined/cli/dev.py
Comment (wes-turner) [inline harness/determined/cli/dev.py:251]: nit; I expect this means "return value", but it's nonstandard for me and a little cryptic. Even just `result` reads more clearly to me.

File: harness/tests/cli/test_cli.py
Comment (wes-turner) [inline harness/tests/cli/test_cli.py:455]: Though this isn't traditionally how we've tested the CLI, we're trying to in general [move more to unit tests](https://hpe-aiatscale.atlassian.net/wiki/spaces/ENGINEERIN/pages/1351909582/Python+Testing). Please let me know if you'd like to chat about it at some point, and no worries if that point isn't anytime soon.

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:10]: this was mainly here to let me `eval` the annotations. I moved it down there. I do like sticking to `typing` and not importing the type hints directly but much more if we alias it to something short same as `numpy as np` otherwise a compound type gets way too long and we need more type aliases and so on.
I updated it

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:71]: > How is it different than `bindings_sig_str`?

the latter returns a user friendly string representation of it

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:109]: supported here is: if our CLI deserializer supports a given type annotation.

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:110]: I don't remember how but there was a related CI failure on our CI 3.7 tests 🤔

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:130]: is this about no new line after the first """ ? or writing a summary? There should be a linter for the former, it seems easy to automate.

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:133]: I like this better too

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:177]: >  Why is it relevant that value is invalid JSON?

eg if u pass `trialIds=23` then the default json decoder error was way less clear to me. maybe contacting them together gives us both

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:86]: That's a good point but IMO this makes it a bit easier to experiment with more complex styles. I prefer to keep it as is keep it as is. let me know if you think it matters much

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:265]: `result` could be a side effect? but return value is more explicit? 🤔  either way a longer name is better I agree. I'll update it

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:266]: TypeError and (since our annotations aren't always accurate or as strict as they could be ) whatever exception our apis throw on error I was just lazy to look up all of those.

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:253]: I agree it's a bit odd, I have thoughts on making it better but it all comes with more code
the way I've been using it is "I know there is an api called similar to 'comparetrial' but I don't know the full name can I just call it without exactly knowing what it was?" and this flow has helped more that erroring out and exiting. IMO If we run into a concrete downside with it (maybe piping?) we can reevaluate. I updated the wording.

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:97]: our annotations are [defined](https://github.com/determined-ai/determined/blob/fa8781bdb564d10a5e998af9dbbfc648015914b5/harness/determined/common/api/bindings.py#L412) as `str`s  as part of an import speed optimization the way I've been able to still use it with the `typing` helpers is through `eval` maybe there is a better way

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py:157]: > are similar

one operates on a full function call the other evaluates the type referred to by each function parameter. can you expand a bit?

File: harness/determined/cli/dev.py
Comment (hamidzr) [reply harness/determined/cli/dev.py]: I think I updated these

Comment (hamidzr) [reply harness/determined/cli/dev.py]: I agree but I don't intend to do more with this but if we want to go that route it might make sense to be housed under `cli.render` and then tested and changed to return a string instead.

File: harness/tests/cli/test_cli.py
Comment (hamidzr) [reply harness/tests/cli/test_cli.py:455]: yeah, this will cause test failures when this particular API changes which isn't good. it's just been a lazy replacement for building out a list of annotations. I'll switch it out.

**Comment by hamidzr** [comment]
@wes-turner thanks for the comments I went ahead and updated the code and tried to address all the ones that were clear and straight forward to act on to me

**Comment by hamidzr** [comment]
(@wes-turner is OOO) I'll land at this stage for now and i'll put up a follow-up PR if needed for any new comments.

**Pull request merged**
